### PR TITLE
feat: support unique credential cookie names

### PR DESCRIPTION
**Pull request by limhjgrace**
Multiple AppMonitors from different AWS accounts cannot currently be used on the same page. Because the credential cookie name `cwr_c` is not unique to an AppMonitor, multiple web clients running on the same page attempt to read and store credential info to the same cookie, which would produce incorrect data.

This change uses the existing `cookieAttributes.unique` configuration to add the AppMonitorID as a postfix to the credential cookie `cwr_c`. When used, the credential cookie name will have the format `cwr_c_[AppMonitor Id]`, and multiple web clients sending data to app monitors in different accounts can run on the same page. 

Related FR: https://github.com/aws-observability/aws-rum-web/issues/557

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ps863** [comment]
question: will we update existing doc/provide examples in separate PR?

**Review by ps863 — APPROVED** [review]

**Comment by ps863** [comment]
question: if there are multiple web clients running for different app monitors, user would have to make sure the unique configuration is enabled in each one for this to work correctly right?

**Comment by limhjgrace** [comment]
> question: will we update existing doc/provide examples in separate PR?

Thanks for the reminder, updated existing doc in latest commit.

**Comment by limhjgrace** [comment]
> question: if there are multiple web clients running for different app monitors, user would have to make sure the unique configuration is enabled in each one for this to work correctly right?

Technically if they had 2 app monitors, only one would need to configure unique cookies because they would read from different cookies. But as best practice, users should configure each app monitor, if using multiple, to configure using unique cookies.

File: src/dispatch/Authentication.ts [hunk 1]
```diff
@@ -3,19 +3,22 @@ import { Config } from '../orchestration/Orchestration';
...
+    protected applicationId: string;
     protected cognitoIdentityClient: CognitoIdentityClient;
     protected config: Config;
     protected credentials: AwsCredentialIdentity | undefined;
 
-    constructor(config: Config) {
+    constructor(config: Config, applicationId: string) {
         const region: string = config.identityPoolId!.split(':')[0];
         this.config = config;
         this.cognitoIdentityClient = new CognitoIdentityClient({
```
Comment (qhanam) [inline src/dispatch/Authentication.ts:17]: issue: We need unique keys for the Cognito identity as well.

File: src/dispatch/Authentication.ts [hunk 2]
```diff
@@ -3,19 +3,22 @@ import { Config } from '../orchestration/Orchestration';
...
             fetchRequestHandler: new FetchHttpHandler(),
             region
         });
+        this.applicationId = applicationId;
```
Comment (qhanam) [inline src/dispatch/Authentication.ts:21]:
```suggestion
        this.credentialStorageKey = this.config.cookieAttributes.unique
            ? `${CRED_KEY}_${this.applicationId}`
            : CRED_KEY;
```

File: src/dispatch/Authentication.ts [hunk 3]
```diff
@@ -80,7 +83,15 @@ export abstract class Authentication {
...
                 try {
-                    credentials = JSON.parse(localStorage.getItem(CRED_KEY)!);
+                    credentials = JSON.parse(
+                        localStorage.getItem(
+                            getCookieName(
+                                this.config.cookieAttributes.unique,
+                                CRED_KEY,
+                                this.applicationId
+                            )
+                        )!
```
Comment (qhanam) [inline src/dispatch/Authentication.ts:93]:
```suggestion
                            this.credentialStorageKey
```

File: src/sessions/SessionManager.ts [hunk 4]
```diff
@@ -176,7 +176,11 @@ export class SessionManager {
     private createOrRenewSessionCookie(session: Session, expires: Date) {
         if (btoa) {
             storeCookie(
-                this.sessionCookieName(),
+                getCookieName(
+                    this.config.cookieAttributes.unique,
+                    SESSION_COOKIE_NAME,
+                    this.appMonitorDetails.id!
+                ),
```
Comment (qhanam) [inline src/sessions/SessionManager.ts:183]: suggestion: Set the cookie name once in the constructor.
```
this.sessionCookieName = this.config.cookieAttributes.unique
            ? `${SESSION_COOKIE_NAME}_${this.appMonitorDetails.id}`
            : SESSION_COOKIE_NAME;
```
Then we can re-use the value.
```suggestion
                this.sessionCookieName,
```

**Comment by qhanam** [comment]
I think this change needs to be integration tested somehow. That is, when unique keys are enabled, and multiple web clients are used concurrently, then authentication succeeds.

For example:
1. When unique keys are enabled, does authentication succeed?
2. When there are multiple app monitor instances, does authentication succeed?

I don't know if mocking the requests to Cognito and STS makes sense here. Mocking might be OK if we at least manually verify authentication succeeds for case (2). Otherwise, these might need to be run in the smoke test environment, which makes remote requests.

**Comment by limhjgrace** [comment]
> I think this change needs to be integration tested somehow. That is, when unique keys are enabled, and multiple web clients are used concurrently, then authentication succeeds.
> 
> For example:
> 
> 1. When unique keys are enabled, does authentication succeed?
> 2. When there are multiple app monitor instances, does authentication succeed?
> 
> I don't know if mocking the requests to Cognito and STS makes sense here. Mocking might be OK if we at least manually verify authentication succeeds for case (2). Otherwise, these might need to be run in the smoke test environment, which makes remote requests.

Added smoke tests for applications using npm. The second web client is created after a 10 second wait from the first one to ensure credentials are retrieved and loaded separately. During smoke test dev, we discovered that we currently don't support multiple web clients for applications using CDN.

**Comment by limhjgrace** [comment]
> > I think this change needs to be integration tested somehow. That is, when unique keys are enabled, and multiple web clients are used concurrently, then authentication succeeds.
> > For example:
> > 
> > 1. When unique keys are enabled, does authentication succeed?
> > 2. When there are multiple app monitor instances, does authentication succeed?
> > 
> > I don't know if mocking the requests to Cognito and STS makes sense here. Mocking might be OK if we at least manually verify authentication succeeds for case (2). Otherwise, these might need to be run in the smoke test environment, which makes remote requests.
> 
> Added smoke tests for applications using npm. The second web client is created after a 10 second wait from the first one to ensure credentials are retrieved and loaded separately. During smoke test dev, we discovered that we currently don't support multiple web clients for applications using CDN.

Smoke tests successfully pass: https://github.com/limhjgrace/aws-rum-web/actions/runs/9487042028/job/26142730870

**Review by williazz — APPROVED** [review]

**Review by qhanam — APPROVED** [review]

**Comment by limhjgrace** [comment]
Smoke tests passed: https://github.com/limhjgrace/aws-rum-web/actions/runs/9847231882/job/27186743456. Need to update secret keys for main repo.

**Pull request merged**